### PR TITLE
Added: new rule `selector-descendant-combinator-no-non-space`.

### PR DESCRIPTION
**Pull request by evilebottnawi**
Ref: https://github.com/stylelint/stylelint/issues/1592
/cc @davidtheclark @jeddy3

### CODE REVIEW

File: docs/user-guide/rules.md [hunk 1]
```diff
@@ -156,6 +156,7 @@ Here are all the rules within stylelint, grouped by the [*thing*](http://apps.wo
 - [`selector-class-pattern`](../../src/rules/selector-class-pattern/README.md): Specify a pattern for class selectors.
 - [`selector-combinator-space-after`](../../src/rules/selector-combinator-space-after/README.md): Require a single space or disallow whitespace after the combinators of selectors.
 - [`selector-combinator-space-before`](../../src/rules/selector-combinator-space-before/README.md): Require a single space or disallow whitespace before the combinators of selectors.
+- [`selector-descendant-combinator-no-non-space`](../../src/rules/selector-descendant-combinator-no-non-space/README.md): Disable non-spaces characters of the descendant combinator in descendant selectors.
```
Comment (m-allanson) [inline docs/user-guide/rules.md]: I think this would be clearer as "Disable non-space characters for the descendant combinator of a selector."

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# selector-descendant-combinator-no-non-space
+
+Disable non-spaces characters of the descendant combinator in descendant selectors.
```
Comment (m-allanson) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: See comment about rules.md

**Comment by m-allanson** [comment]
A minor comment about the description, otherwise LGTM 👍

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+## Options
+
+### `true`
+
+The following patterns are considered warnings:
+
+```css
+.selector1 
+.selector2 {}
+```
```
Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: A little more explanation is needed because "non-space" is not a self-evident term.

File: src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  }, {
+    code: ".selector1  .selector2 {}",
+  }, {
+    code: ".selector1>.selector2 {}",
+  }, {
+    code: ".selector1 > .selector2 {}",
+  }, {
+    code: ".selector1  >  .selector2 {}",
+  }, {
+    code: ".selector1\n>\n.selector2 {}",
```
Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js]: Shouldn't this and the following be a warning?

File: src/rules/selector-descendant-combinator-no-non-space/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+import { nonSpaceCombinators } from "../../reference/punctuationSets"
+
+export const ruleName = "selector-descendant-combinator-no-non-space"
+
+export const messages = ruleMessages(ruleName, {
+  rejected: nonSpaceCharacter => `Unexpected non-space "${nonSpaceCharacter}" characters in descendant combinator`,
```
Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/index.js]: How about just `Unexpected "${nonSpaceCharacter}" in descendant combinator`

File: src/rules/selector-descendant-combinator-no-non-space/index.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+export default function (actual) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, { actual })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      if (!isStandardSyntaxRule(rule)) { return }
+
+      const { selector } = rule
+      if (!isStandardSyntaxSelector(selector)) { return }
```
Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/index.js]: Why is this necessary?

File: docs/user-guide/rules.md (hunk 1 above)
Comment (jeddy3) [inline docs/user-guide/rules.md]: How's about?

"Disallow non-space characters for descendant combinators of selectors."

As we use "disallow" and plurals elsewhere.

File: src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  }, {
+    code: ".selector1 .selector2 {}",
+  }, {
+    code: ".selector1  .selector2 {}",
+  }, {
+    code: ".selector1>.selector2 {}",
+  }, {
+    code: ".selector1 > .selector2 {}",
+  }, {
+    code: ".selector1  >  .selector2 {}",
```
Comment (jeddy3) [inline src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js]: We've used `space` in the rule name elsewhere to consistently means *a single space*. I think this rule should behave in the same way.

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+The following patterns are *not* considered warnings:
+
+```css
+.selector1 .selector2 {}
+```
+
+```css
+.selector1  .selector2 {}
```
Comment (jeddy3) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: See https://github.com/stylelint/stylelint/pull/1784/files#r74931454

File: src/rules/selector-descendant-combinator-no-non-space/index.js (hunk 5 above)
Comment (jeddy3) [inline src/rules/selector-descendant-combinator-no-non-space/index.js]: We've removed redundancy between the message and the rule name elsewhere, as such I think we could even shorten this to `Unexpected "${nonSpaceCharacter}"`

e.g.

`Unexpected "\t" (selector-descendant-combinator-no-non-space)`

Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/index.js]: 👍

File: src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js (hunk 7 above)
Comment (evilebottnawi) [inline src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js]: @jeddy3 agree, but this test should accepted, because it is not `descendant-combinator`.

Comment (jeddy3) [inline src/rules/selector-descendant-combinator-no-non-space/__tests__/index.js]: Right you are.

It's the test three above, i.e. `.selector1  .selector2 {}`, that should be rejected.

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# selector-descendant-combinator-no-non-space
+
+Disable non-spaces characters of the descendant combinator in descendant selectors.
+
+```css
+   .selector1 .selector2 .selector3 { color: pink; }
```
Comment (jeddy3) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: Can you update these (and the patterns below) to `.foo`, `.bar` and `.baz` for consistency with the other READMEs?

Comment (evilebottnawi) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: @jeddy3 yep

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
+# selector-descendant-combinator-no-non-space
+
+Disallow non-space characters for descendant combinators of selectors.
+
+```css
+   .foo .bar .baz { color: pink; }
+/**          ↑          ↑
```
Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: Arrows are off.

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * These are descendant combinators */
+```
+
+This rule ensures that there is only single space for descendant combinators of selectors.
+
```
Comment (davidtheclark) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: Please somewhere in this description include explicit mention of tabs, newlines, and multiple spaces. I still don't think the purpose of this rule is made immediately obvious by these docs.

**Comment by davidtheclark** [comment]
I'm fine with this other than the need for more explanatory documentation.

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+Disallow non-space characters for descendant combinators of selectors.
+
+```css
+   .foo .bar .baz { color: pink; }
+/**    ↑         ↑
+ * These are descendant combinators */
+```
+
+This rule ensures that there is only single space are used and verifies that tabs, newlines, and  multiple spaces are not used for descendant combinators of selectors.
```
Comment (ntwb) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: `This rule ensures that only a single space is used and ensures no tabs, newlines, nor multiple spaces are used for descendant combinators of selectors.`

That's my take on more succinct verbiage?

Comment (evilebottnawi) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: @ntwb thanks, and sorry for my bad english :smile:

File: src/rules/selector-descendant-combinator-no-non-space/README.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
+# selector-descendant-combinator-no-non-space
+
+Disallow non-space characters for descendant combinators of selectors.
+
```
Comment (jeddy3) [inline src/rules/selector-descendant-combinator-no-non-space/README.md]: Please update to:

```css
.foo .bar .baz { color: pink; }
/** ↑    ↑
* These descendant combinators */
```

So the arrows point to the right place.

**Comment by jeddy3** [comment]
@davidtheclark This is almost finished. Shall I make the final tweaks and get it in?

**Comment by davidtheclark** [comment]
> Shall I make the final tweaks and get it in?

That would be great. I actually was just trying to do the same with https://github.com/stylelint/stylelint/pull/1792.
